README.md
<h1 align="center">Welcome to react-native-paginated-modal-picker 👋</h1>
<p>
  <a href="https://www.npmjs.com/package/react-native-paginated-modal-picker" target="_blank">
    <img alt="Version" src="https://img.shields.io/npm/v/react-native-paginated-modal-picker.svg">
  </a>
  <a href="#" target="_blank">
    <img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg" />
  </a>
</p>

> Single and Multiple React Native Modal Picker With Pagination!

## Install

```sh
npm i react-native-paginated-modal-picker
```

https://user-images.githubusercontent.com/40437835/153620306-3d4b6ceb-ac96-40bd-80b6-2889281ef0ec.mov

## Single and Multi Picker Usage

```jsx
import React, { useState, useEffect } from 'react';
import { View, Text, Dimensions, Image } from 'react-native';
import {
  PaginatedModalPicker,
  PaginatedMultipleModalPicker,
} from 'react-native-paginated-modal-picker';
import axios from 'axios';
const { width, height } = Dimensions.get('window');
export default function App() {
  const [users, setUsers] = useState({
    data: [],
    pageNumber: 1,
    isLoading: false,
  });

  const pages = 2;
  const [selectedUser, setSelectedUser] = useState({
    name: '',
    id: '',
  });

  const [selectedUsers, setSelectedUsers] = useState({
    names: [],
    ids: [],
  });

  const getCountriesData = async (page) => {
    const response = await axios.get(
      `https://reqres.in/api/users?page=${page}`,
      {
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    return response.data.data;
  };

  useEffect(() => {
    getCountriesData(users.pageNumber).then((response) => {
      const usersLabeled = response.map((data) =>
        Object.assign({}, { label: data.first_name, value: data.id })
      );
      setUsers({
        ...users,
        data: [...users.data, ...usersLabeled],
        isLoading: false,
      });
    });
  }, [users.pageNumber]);

  return (
    <View
      style={{ flex: 1, alignItems: 'center', justifyContent: 'space-around' }}>
      <PaginatedModalPicker
        textInputStyle={{ paddingLeft: 5 }}
        // modalStyle={{ height: height * 0.4 }}
        flatListStyle={{ padding: 6 }}
        listItemStyle={{
          borderRadius: 5,
          borderColor: 'black',
          borderWidth: 0.4,
          marginTop: 20,
        }}
        listItemTextStyle={{ color: 'blue' }}
        placeholderTextColor='#9b9b9b'
        placeholder={'Select a User'}
        isRequestingComponent={<Text>Loading Component</Text>}
        ListEmptyComponent={<Text>Empty Component</Text>}
        isWorking={true}
        data={users.data}
        // modalType={'FullScreen'}
        backDropOpacity={0.6}
        onSelect={(item) => {
          setSelectedUser({ id: item.value, name: item.label });
        }}
        value={selectedUser.name}
        onEndReached={() =>
          pages > users.pageNumber &&
          setUsers({
            ...users,
            pageNumber: users.pageNumber + 1,
            isLoading: true,
          })
        }
        isRequesting={users.isLoading}
        onEndReachedThreshold={0}
      />

      <PaginatedMultipleModalPicker
        textInputStyle={{ paddingLeft: 5 }}
        // modalStyle={{ height: height * 0.4 }}
        flatListStyle={{ padding: 6 }}
        listItemStyle={{
          borderRadius: 5,
          borderColor: 'black',
          borderWidth: 0.4,
          marginTop: 20,
        }}
        listItemTextStyle={{ color: 'blue' }}
        //new props start
        saveButtonStyle={{ backgroundColor: '#1259a1' }}
        saveButtonTitleStyle={{ fontSize: 17 }}
        saveButtonTitle='Confirm'
        selectedItemImageStyle={{ width: 13, height: 13 }}
        selectedItemIconComponent={
          <Image
            source={require('./images/correct.png')}
            style={{ height: 20, width: 20 }}
          />
        }
        //new props end
        placeholderTextColor='#9b9b9b'
        placeholder={'Select Multiple Users'}
        isRequestingComponent={<Text>Loading Component</Text>}
        ListEmptyComponent={<Text>Empty Component</Text>}
        isWorking={true}
        data={users.data}
        // modalType={'FullScreen'}
        backDropOpacity={0.6}
        onSelect={(item) => {
          setSelectedUsers({ ids: item.values, names: item.labels });
        }}
        value={selectedUsers?.names?.toString()}
        onEndReached={() =>
          pages > users.pageNumber &&
          setUsers({
            ...users,
            pageNumber: users.pageNumber + 1,
            isLoading: true,
          })
        }
        isRequesting={users.isLoading}
        onEndReachedThreshold={0}
      />
    </View>
  );
}
```

## Properties

### Single Picker

#### Props

| Prop                  | Default                  | Type                 | Description                                                                                     |
| --------------------- | ------------------------ | -------------------- | ----------------------------------------------------------------------------------------------- |
| data                  | `[{label:'', value:''}]` | `array`              | an array pagination items                                                                       |
| placeholder           | `Click Me!`              | `string`             | place holder for textinput                                                                      |
| isWorking             | `true`                   | `bool`               | enable/disable textinput                                                                        |
| isRequesting          | `false`                  | `bool`               | shows flatlist footer component when requesting data with pagination                            |
| onSelect              | ` `                      | `call back function` | used to get the selected label and value                                                        |
| value                 | ` `                      | `string`             | value of text input after onSelect is fired                                                     |
| modalType             | `Custom`                 | `string`             | when modalType is 'FullScreen' it become in fullscreen view                                     |
| onEndReached          | ` `                      | `function`           | triggered when end of flatlist reached incase of pagination                                     |
| onEndReachedThreshold | `number`                 | `number`             | represents the number of screen lengths you should be from the bottom before it fires the event |

#### Components

| Components            | Description         |
| --------------------- | ------------------- |
| ListEmptyComponent    | ListEmptyComponent  |
| isRequestingComponent | ListFooterComponent |

#### Styles

| Prop                  | Default                                                                                                        | Type          | Description                        |
| --------------------- | -------------------------------------------------------------------------------------------------------------- | ------------- | ---------------------------------- |
| textInputStyle        | `{ borderRadius: 10, borderWidth: 1, width: width * 0.8, height: height * 0.05, padding: 10}`                  | `style`       | textInputStyle                     |
| modalStyle            | `{ borderRadius: 10, width: width * 0.8, height: height * 0.3, backgroundColor: 'white', alignSelf: 'center'}` | `style`       | modalStyle                         |
| contentContainerStyle | `{}`                                                                                                           | `style`       | contentContainerStyle for flatlist |
| flatListStyle         | `{}`                                                                                                           | `style`       | style for flatlist                 |
| placeholderTextColor  | ` `                                                                                                            | `hexadecimal` | placeholderTextColor               |
| listItemStyle         | `{padding: 10}`                                                                                                | `style`       | style for each item in flatlist    |
| listItemTextStyle     | `{fontSize: 16}`                                                                                               | `style`       | listItemTextStyle                  |
| backDropOpacity       | `0.65`                                                                                                         | `number`      | backDropOpacity of modal           |

### Multi Picker ( Can use same Props, Components and Styles of Single Picker in addition to: )

#### Props

| Prop            | Default | Type     | Description          |
| --------------- | ------- | -------- | -------------------- |
| saveButtonTitle | `Save`  | `string` | title of save button |

#### Styles

| Prop                  | Default                                                                                                                                                                                                           | Type    | Description           |
| --------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | --------------------- |
| saveButtonStyle       | `{borderRadius: 30, minHeight: height * 0.04, width: width * 0.3, backgroundColor: 'black', justifyContent: "center", alignItems: 'center', alignSelf: "center", marginTop: 20, marginHorizontal: width * 0.015}` | `style` | saveButtonStyle       |
| saveButtonTitleStyle  | `{color: "white", fontSize: width * 0.04}`                                                                                                                                                                        | `style` | saveButtonTitleStyle  |
| selectedItemIconStyle | `{height: 15, width: 15, alignSelf: 'center'}`                                                                                                                                                                    | `style` | selectedItemIconStyle |

#### Components

| Components                | Description               |
| ------------------------- | ------------------------- |
| selectedItemIconComponent | selectedItemIconComponent |

## Author

👤 **Ahmed Halbas**

- Github: [@AhmedHalbas](https://github.com/AhmedHalbas)
- LinkedIn: [@AhmedHalbas](https://linkedin.com/in/AhmedHalbas)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/AhmedHalbas/react-native-paginated-modal-picker/issues).

## Show your support

Give a ⭐️ if this project helped you!


CHANGELOG.md
# Changelog

### 1.1.0 (January 14, 2022)

- Improve Code Quality.
- Add New Props to Paginated Modal: `value`, `modalType`, `backDropOpacity`.

### 2.0.0 (February 11, 2022)

- Add `PaginatedMultipleModalPicker`.
- With New Props: `saveButtonStyle`, `saveButtonTitleStyle`, `saveButtonTitle`, `selectedItemIconStyle`, `selectedItemIconComponent`.

### 2.0.1 (July 27, 2022)

- Handle Initial Values in `data` prop in `PaginatedMultipleModalPicker`.
- Improve Code Quality.

### 2.0.2 (December 08, 2022)

- Improve Package Code Quality.


Package/Components/paginatedMultipleModalPicker.js
import React, { useState, useEffect } from 'react';
import {
    View,
    Platform,
    FlatList,
    TouchableOpacity,
    ActivityIndicator,
    Modal,
    Text,
    Dimensions,
    TextInput,
    Image
} from 'react-native';
import useUpdateEffect from '../Helpers/useUpdateEffectHook'
const { width, height } = Dimensions.get("window");

export default function PaginatedMultipleModalPicker({ value, modalType = 'Custom', isWorking = true, placeholder = 'Click Me!', data, onSelect, onEndReached, onEndReachedThreshold, isRequesting, textInputStyle, modalStyle, backDropOpacity, contentContainerStyle, flatListStyle, placeholderTextColor, listItemStyle, listItemTextStyle, saveButtonStyle, saveButtonTitleStyle, saveButtonTitle, selectedItemIconStyle, ListEmptyComponent, isRequestingComponent, selectedItemIconComponent }) {

    const [isVisible, setVisible] = useState(false);
    const [onScrollEnd, setOnScrollEnd] = useState(false);

    const [dataItems, setDataItems] = useState([]);


    const isModalFullScreen = () => modalType == 'FullScreen';

    useEffect(() => {

        setDataItems(data)

    }, [data])


    useUpdateEffect(() => {

        if (Array.isArray(value) && value.length) {
            let currentData = data
            const selectedLabels = value.split(',')
            for (let item of currentData) {
                if (selectedLabels.includes(item.label)) {
                    item.selected = true
                }
            }
            setDataItems(currentData)
        }
    }, [value])


    const selectItem = (id) => {

        let currentData = [...dataItems];

        for (let data of currentData) {
            if (data.value == id) {
                data.selected = (data.selected == null) ? true : !data.selected;
                break;
            }
        }
        setDataItems(currentData)

    }

    const getSelected = (key) => dataItems.filter(item => item.selected == true).map(itm => itm[key])

    const handleOnSelect = () => {
        setVisible(false);

        const selectedObj = {
            values: getSelected('value')
            , labels: getSelected('label')
        }
        onSelect?.(selectedObj);

    }


    return (
        <View>

            <TouchableOpacity disabled={!isWorking} onPress={() => setVisible(true)} >
                <View pointerEvents='none'>

                    <TextInput
                        style={{ borderRadius: 10, borderWidth: 1, width: width * 0.8, height: height * 0.05, padding: 10, ...textInputStyle }}
                        editable={false}
                        placeholder={placeholder}
                        value={value}
                        placeholderTextColor={placeholderTextColor}
                    />
                </View>

            </TouchableOpacity>

            <Modal

                visible={isVisible}
            >

                <TouchableOpacity
                    activeOpacity={0}
                    style={{
                        flex: 1,
                        justifyContent: 'center',
                        alignItems: 'center',
                        backgroundColor: 'black',
                        opacity: backDropOpacity ?? 0.65
                    }}
                    onPress={handleOnSelect}>
                    <TouchableOpacity activeOpacity={1} >
                        <View style={{ backgroundColor: 'white', borderRadius: 10, paddingTop: isModalFullScreen() && Platform.OS == 'ios' ? height * 0.06 : 0, width: isModalFullScreen() ? width : width * 0.8, height: isModalFullScreen() ? height : height * 0.3, alignSelf: 'center', ...modalStyle }}>
                            <FlatList
                                ListHeaderComponent={isModalFullScreen() && <TouchableOpacity
                                    onPress={() => setVisible(false)}
                                    style={{
                                        width: 30,
                                        height: 30,
                                        justifyContent: 'center',
                                        alignItems: 'center'

                                    }}>
                                    <Text style={{ fontSize: 20 }}>✕</Text>
                                </TouchableOpacity>
                                }
                                style={{ padding: 10, ...flatListStyle }}
                                contentContainerStyle={{ paddingBottom: height * 0.05, ...contentContainerStyle }}
                                ListEmptyComponent={
                                    ListEmptyComponent ??
                                    <View style={{ justifyContent: 'center', alignSelf: 'center', flex: 1 }}>
                                        <Text style={{ fontSize: 14, textAlign: 'center' }}>
                                            No Content Found
                                        </Text>
                                    </View>
                                }
                                ListFooterComponent={
                                    <>
                                        {
                                            isRequesting ?
                                                (
                                                    isRequestingComponent ? (isRequestingComponent) :
                                                        (
                                                            <View>
                                                                <ActivityIndicator size="large" />
                                                            </View>

                                                        )
                                                ) :

                                                (
                                                    <View
                                                    >
                                                        <TouchableOpacity
                                                            style={{ borderRadius: 30, minHeight: height * 0.04, width: width * 0.3, backgroundColor: 'black', justifyContent: "center", alignItems: 'center', alignSelf: "center", marginTop: 20, marginHorizontal: width * 0.015, ...saveButtonStyle }}
                                                            onPress={() => {
                                                                if (dataItems.filter(item => item.selected == true)?.length > 0) {
                                                                    handleOnSelect()
                                                                }

                                                            }}

                                                        >
                                                            <Text style={{ color: "white", fontSize: width * 0.04, ...saveButtonTitleStyle }}>{saveButtonTitle ?? 'Save'}</Text>
                                                        </TouchableOpacity>
                                                    </View>
                                                )

                                        }

                                    </>

                                }

                                data={dataItems}
                                renderItem={(data) => {
                                    return (
                                        <TouchableOpacity
                                            style={{ padding: 10, ...listItemStyle }}
                                            onPress={() => selectItem(data.item.value)}>


                                            <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
                                                <Text style={{ fontSize: 16, ...listItemTextStyle }}>{data.item.label}</Text>

                                                {data.item.selected && (selectedItemIconComponent ? selectedItemIconComponent : <Image
                                                    source={require('../Images/check.png')}
                                                    style={{ height: 15, width: 15, alignSelf: 'center', ...selectedItemIconStyle }}
                                                />)}
                                            </View>

                                        </TouchableOpacity>
                                    );
                                }}
                                keyExtractor={(item, index) => index.toString()}
                                onEndReached={() => onEndReached && setOnScrollEnd(true)}
                                onMomentumScrollEnd={() => {
                                    onScrollEnd && onEndReached?.()
                                    setOnScrollEnd(false)
                                }}
                                onEndReachedThreshold={onEndReachedThreshold ?? (Platform.OS == 'ios') ? 0 : 3}

                            />
                        </View>
                    </TouchableOpacity>
                </TouchableOpacity>

            </Modal>
        </View>
    );

}